### PR TITLE
IHP130 OpenPDK support for ORFS

### PR DESCRIPTION
**Pull request by KrzysztofHerman**
This is the first version of the official IHP 130 OpenPDK support for ORFS. 
The repository still lacks IO cells, LVS rules and the DRC ruleset is in the preliminary version.
In the Makefile one can find the designs which were successfully synthesized using ORFS. 
Some minor changes were made in `utils/def2stream.py` in order to support correct streaming of the final GDS using 
klayout.

### CODE REVIEW

File: flow/Makefile [hunk 1]
```diff
@@ -94,9 +94,15 @@
 # DESIGN_CONFIG=./designs/gf180/jpeg/config.mk
 # DESIGN_CONFIG=./designs/gf180/riscv32i/config.mk
 # DESIGN_CONFIG=./designs/gf180/uart-blocks/config.mk
-#
+
+#DESIGN_CONFIG=./designs/ihp-sg13g2/aes/config.mk
+#DESIGN_CONFIG=./designs/ihp-sg13g2/ibex/config.mk
+DESIGN_CONFIG=./designs/ihp-sg13g2/gcd/config.mk
```
Comment (maliberty) [inline flow/Makefile]: This should be commented out

File: flow/Makefile [hunk 2]
```diff
@@ -94,9 +94,15 @@
...
+#DESIGN_CONFIG=./designs/ihp-sg13g2/spi/config.mk
+#DESIGN_CONFIG=./designs/ihp-sg13g2/riscv32i/config.mk
+
 # Default design
-DESIGN_CONFIG ?= ./designs/nangate45/gcd/config.mk
+#DESIGN_CONFIG ?= ./designs/ihp-sg13g2/gcd/config.mk
```
Comment (maliberty) [inline flow/Makefile]: Don't change the default

File: flow/Makefile [hunk 3]
```diff
@@ -1076,3 +1082,7 @@ klayout:
 include $(UTILS_DIR)/utils.mk
 export PRIVATE_DIR ?= ../../private_tool_scripts
 -include $(PRIVATE_DIR)/private.mk
+
+
+
+
```
Comment (maliberty) [inline flow/Makefile]: rm

File: flow/designs/ihp-sg13g2/aes/config.mk [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+export CORE_MARGIN = 2
+
+export PLACE_DENSITY = 0.65
+export TNS_END_PERCENT = 100
+
+export USE_FILL = 1
+
+#export PLACE_DENSITY ?= 0.88
+#export CORE_UTILIZATION = 20
+#export TNS_END_PERCENT = 100
```
Comment (maliberty) [inline flow/designs/ihp-sg13g2/aes/config.mk]: rm

File: flow/designs/ihp-sg13g2/gcd/config.mk [hunk 5]
```diff
@@ -0,0 +1,13 @@
+export DESIGN_NAME = gcd
+export PLATFORM    = ihp-sg13g2
+
+export VERILOG_FILES = ./designs/src/$(DESIGN_NICKNAME)/gcd.v
+export SDC_FILE      = ./designs/$(PLATFORM)/$(DESIGN_NICKNAME)/constraint.sdc
+
+# Adders degrade GCD
```
Comment (maliberty) [inline flow/designs/ihp-sg13g2/gcd/config.mk]: Is this true in this pdk?

File: flow/designs/ihp-sg13g2/spi/config.mk [hunk 6]
```diff
@@ -0,0 +1,13 @@
+export DESIGN_NAME = spi
+export PLATFORM    = ihp-sg13g2
+
+export VERILOG_FILES = ./designs/$(PLATFORM)/$(DESIGN_NICKNAME)/src/spi.v
+export SDC_FILE      = ./designs/$(PLATFORM)/$(DESIGN_NICKNAME)/spi.sdc
+
+# Adders degrade GCD
+export ADDER_MAP_FILE :=
```
Comment (maliberty) [inline flow/designs/ihp-sg13g2/spi/config.mk]: Not gcd

File: flow/designs/ihp-sg13g2/spi/spi.sdc [hunk 7]
```diff
@@ -0,0 +1,9 @@
+###################################################################
+
+# Created by write_sdc on Fri Jun 20 08:42:53 2014
+
+###################################################################
+set sdc_version 1.9
+
+create_clock [get_ports clk]  -period 20  -waveform {0 10}
+set_clock_uncertainty 2  [get_clocks clk]
```
Comment (maliberty) [inline flow/designs/ihp-sg13g2/spi/spi.sdc]: No timing on the inputs & outputs?

File: flow/designs/ihp-sg13g2/spi/src/spi.v
Comment (maliberty) [inline flow/designs/ihp-sg13g2/spi/src/spi.v:1]: Should be in flow/designs/src/spi

Comment (maliberty) [reply flow/designs/ihp-sg13g2/spi/src/spi.v:1]: What is this design and what does it provide?

File: flow/platforms/ihp-sg13g2/cells_adders.v [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  
+  parameter _TECHMAP_CONSTVAL_A_ = WIDTH'bx;
+  parameter _TECHMAP_CONSTVAL_B_ = WIDTH'bx;
+  parameter _TECHMAP_CONSTVAL_C_ = WIDTH'bx;
+  
+  genvar i;
+  generate for (i = 0; i < WIDTH; i = i + 1) begin
+      if (_TECHMAP_CONSTVAL_A_[i] === 1'b0 || _TECHMAP_CONSTVAL_B_[i] === 1'b0 || _TECHMAP_CONSTVAL_C_[i] === 1'b0) begin
+        if (_TECHMAP_CONSTVAL_C_[i] === 1'b0) begin
+          sky130_fd_sc_hd__ha_1 halfadder_Cconst (
```
Comment (maliberty) [inline flow/platforms/ihp-sg13g2/cells_adders.v]: This can't be right as you are using cells for sky130

File: flow/platforms/ihp-sg13g2/cells_latch.v [hunk 9]
```diff
@@ -0,0 +1,15 @@
+module $_DLATCH_P_(input E, input D, output Q);
+    sky130_fd_sc_hd__dlxtp_1 _TECHMAP_REPLACE_ (
```
Comment (maliberty) [inline flow/platforms/ihp-sg13g2/cells_latch.v]: sky130 cells

File: flow/platforms/ihp-sg13g2/config.mk [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+# Tech/Libs
+# ----------------------------------------------------
+export TECH_LEF = $(PLATFORM_DIR)/lef/sg13g2_tech.lef
+export SC_LEF = $(PLATFORM_DIR)/lef/sg13g2_stdcell.lef
+
+export LIB_FILES = $(PLATFORM_DIR)/lib/sg13g2_stdcell_typ_1p20V_25C.lib \
+	                  $(ADDITIONAL_LIBS)
+export GDS_FILES = $(PLATFORM_DIR)/gds/sg13g2_stdcell.gds \
+	                  $(ADDITIONAL_GDS)
+#                     $(ADDITIONAL_GDS)
```
Comment (maliberty) [inline flow/platforms/ihp-sg13g2/config.mk]: rm

File: flow/platforms/ihp-sg13g2/config.mk [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+sg13g2_decap_4 \
+sg13g2_decap_8 \
+sg13g2_lgcp_1 \
+sg13g2_sighold \
+sg13g2_slgcp_1 \
+sg13g2_dfrbp_2 \
+sg13g2_fill_1 \
+sg13g2_fill_2 \
+sg13g2_fill_4 \
+sg13g2_fill_8
```
Comment (maliberty) [inline flow/platforms/ihp-sg13g2/config.mk]: why do fill & decap cells need to be in the dont use list?

File: flow/platforms/ihp-sg13g2/config.mk [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+# This can be found in the technology lef
+export PLACE_SITE = CoreSite
+
+# IO Placer pin layers
+export IO_PLACER_H = Metal2
+export IO_PLACER_V = Metal3
+
+# Define default PDN config
+export PDN_TCL ?= $(PLATFORM_DIR)/pdn.tcl
+#export PDN_TCL ?= $(PLATFORM_DIR)/pdn_grid.cfg
```
Comment (maliberty) [inline flow/platforms/ihp-sg13g2/config.mk]: rm

File: flow/platforms/ihp-sg13g2/config.mk [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+export CELL_PAD_IN_SITES_GLOBAL_PLACEMENT ?= 2
+export CELL_PAD_IN_SITES_DETAIL_PLACEMENT ?= 1
+
+export PLACE_DENSITY ?= 0.65
+
+# --------------------------------------------------------
+#  CTS
+#  -------------------------------------------------------
+# TritonCTS options
+export CTS_BUF_CELL   ?= sg13g2_buf_2
```
Comment (maliberty) [inline flow/platforms/ihp-sg13g2/config.mk:80]: does this pdk have a clock buffer cell?

File: flow/platforms/ihp-sg13g2/config.mk [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+
+# ---------------------------------------------------------
+#  Route
+# ---------------------------------------------------------
+# FastRoute options
+export MIN_ROUTING_LAYER 		?= Metal2
+export MAX_ROUTING_LAYER 		?= Metal5
+#export VIA_IN_PIN_MIN_LAYER ?= Metal1
+#export VIA_IN_PIN_MAX_LAYER ?= Metal1
+#export DISABLE_VIA_GEN      ?= 1
```
Comment (maliberty) [inline flow/platforms/ihp-sg13g2/config.mk:89]: rm

File: flow/platforms/ihp-sg13g2/klayoutrc
Comment (maliberty) [inline flow/platforms/ihp-sg13g2/klayoutrc:1]: What is this file for?

File: flow/platforms/ihp-sg13g2/pdn.tcl [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+#add_global_connection -net {VSS} -inst_pattern {.*} -pin_pattern {^VSSE$}
+#add_global_connection -net {VSS} -inst_pattern {.*} -pin_pattern {VGND}
+#add_global_connection -net {VSS} -inst_pattern {.*} -pin_pattern {VNB}
+####################################
+####################################
+# voltage domains
+####################################
+set_voltage_domain -name {CORE} -power {VDD} -ground {VSS}
+#proc####################################
+# standard c#ell grid
```
Comment (maliberty) [inline flow/platforms/ihp-sg13g2/pdn.tcl]: c#ell -> cell

File: flow/platforms/ihp-sg13g2/pdn.tcl [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+global_connect
+#add_global_connection -net {VDD} -inst_pattern {.*} -pin_pattern {^VDD$} -power
+#add_global_connection -net {VDD} -inst_pattern {.*} -pin_pattern {^VDDPE$}
+#add_global_connection -net {VDD} -inst_pattern {.*} -pin_pattern {^VDDCE$}
+#add_global_connection -net {VDD} -inst_pattern {.*} -pin_pattern {VPWR}
+#add_global_connection -net {VDD} -inst_pattern {.*} -pin_pattern {VPB}
+#add_global_connection -net {VSS} -inst_pattern {.*} -pin_pattern {^VSS$} -ground
+#add_global_connection -net {VSS} -inst_pattern {.*} -pin_pattern {^VSSE$}
+#add_global_connection -net {VSS} -inst_pattern {.*} -pin_pattern {VGND}
+#add_global_connection -net {VSS} -inst_pattern {.*} -pin_pattern {VNB}
```
Comment (maliberty) [inline flow/platforms/ihp-sg13g2/pdn.tcl]: rm commented lines

File: flow/platforms/ihp-sg13g2/pdn_grid.cfg
Comment (maliberty) [inline flow/platforms/ihp-sg13g2/pdn_grid.cfg:1]: when would this file be used?

File: flow/platforms/ihp-sg13g2/setRC.tcl [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+# end correlate
+
+set_layer_rc -via Via1    -resistance 2.0E-3
+set_layer_rc -via Via2    -resistance 2.0E-3
+set_layer_rc -via Via3    -resistance 2.0E-3
+set_layer_rc -via Via4    -resistance 2.0E-3
+set_layer_rc -via TopVia1 -resistance 0.4E-3
+set_layer_rc -via TopVia2 -resistance 0.22E-3
+
+set_wire_rc -signal -layer Metal1
```
Comment (maliberty) [inline flow/platforms/ihp-sg13g2/setRC.tcl]: Metal1 is a pessimistic assumption as you have Metal2 as the min routing layer.

File: flow/util/def2stream.py
Comment (maliberty) [inline flow/util/def2stream.py:1]: Why is this file modified?  Please explain the need to change it for this PDK.

**Review by maliberty — CHANGES_REQUESTED** [review]
Thanks very much for this contribution.

**Comment by maliberty** [comment]
I do find the cells in the library a bit concerning.  For example:
image

1) the clk pin is very small and nearly covered by the metal2 reset shape.
2) the blockages seem to have a lot of extraneous detail.  This is true on almost all cells.  This will slow down the router dealing with the extra complexity.

I wouldn't hold up the PR for this but I think it is something to look at going forward.

File: flow/Makefile (hunk 1 above)
Comment (KrzysztofHerman) [reply flow/Makefile]: Done

File: flow/Makefile (hunk 2 above)
Comment (KrzysztofHerman) [reply flow/Makefile]: Done

File: flow/Makefile (hunk 3 above)
Comment (KrzysztofHerman) [reply flow/Makefile]: Done

File: flow/designs/ihp-sg13g2/aes/config.mk (hunk 4 above)
Comment (KrzysztofHerman) [reply flow/designs/ihp-sg13g2/aes/config.mk]: Done

File: flow/designs/ihp-sg13g2/gcd/config.mk (hunk 5 above)
Comment (KrzysztofHerman) [reply flow/designs/ihp-sg13g2/gcd/config.mk]: We do not provide adders cells so I do not think it is relevant 
```
# Adders degrade GCD
export ADDER_MAP_FILE :=
```
removed

File: flow/platforms/ihp-sg13g2/pdn.tcl (hunk 15 above)
Comment (KrzysztofHerman) [reply flow/platforms/ihp-sg13g2/pdn.tcl]: Done

File: flow/platforms/ihp-sg13g2/pdn.tcl (hunk 16 above)
Comment (KrzysztofHerman) [reply flow/platforms/ihp-sg13g2/pdn.tcl]: Done

File: flow/platforms/ihp-sg13g2/pdn_grid.cfg
Comment (KrzysztofHerman) [reply flow/platforms/ihp-sg13g2/pdn_grid.cfg:1]: It was used for other designs for testing. Removed

File: flow/platforms/ihp-sg13g2/setRC.tcl (hunk 17 above)
Comment (KrzysztofHerman) [reply flow/platforms/ihp-sg13g2/setRC.tcl]: Updated to Metal2

File: flow/util/def2stream.py
Comment (KrzysztofHerman) [reply flow/util/def2stream.py:1]: So the changes in this file were proposed by Matthias Koefferlein, the developer of klayout after discussion about the following error I had enabling the USE_FILL macro. 
```
[INFO][FLOW] Using platform directory ./platforms/ihp-sg13g2
cp ./results/ihp-sg13g2/gcd/base/5_route.sdc ./results/ihp-sg13g2/gcd/base/6_final.sdc
make[1]: Leaving directory '/home/herman/asic/digital/OpenROAD-flow-scripts/flow'
sed '/OR_DEFAULT/d' platforms/ihp-sg13g2/lef/sg13g2_tech.lef > objects/ihp-sg13g2/gcd/base/klayout_tech.lef
sed 's,<lef-files>.*</lef-files>,<lef-files>/home/herman/asic/digital/OpenROAD-flow-scripts/flow/objects/ihp-sg13g2/gcd/base/klayout_tech.lef</lef-files> <lef-files>/home/herman/asic/digital/OpenROAD-flow-scripts/flow/platforms/ihp-sg13g2/lef/sg13g2_stdcell.lef</lef-files>,g' platforms/ihp-sg13g2/sg13g2.lyt > objects/ihp-sg13g2/gcd/base/klayout.lyt
(/usr/bin/time -f 'Elapsed time: %E[h:]min:sec. CPU time: user %U sys %S (%P). Peak memory: %MKB.' stdbuf -o L /bin/klayout -zz -rd design_name=gcd \
        -rd in_def=results/ihp-sg13g2/gcd/base/6_final.def \
        -rd in_files="./platforms/ihp-sg13g2/gds/sg13g2_stdcell.gds  " \
        -rd config_file=./platforms/ihp-sg13g2/fill.json \
        -rd seal_file="" \
        -rd out_file=results/ihp-sg13g2/gcd/base/6_1_merged.gds \
        -rd tech_file=./objects/ihp-sg13g2/gcd/base/klayout.lyt \
        -rd layer_map= \
        -r ./util/def2stream.py) 2>&1 | tee ./logs/ihp-sg13g2/gcd/base/6_1_merge.log
[INFO] Reporting cells prior to loading DEF ...
[INFO] Reading DEF  ...
[INFO] Reporting cells after loading DEF ...
[INFO] 'gcd'
[INFO] 'sg13g2_fill_8'
[INFO] 'sg13g2_fill_4'
[INFO] 'sg13g2_fill_2'
[INFO] 'sg13g2_fill_1'
[INFO] 'sg13g2_nor2b_1'
[INFO] 'sg13g2_nor2b_2'
[INFO] 'sg13g2_nor2_2'
[INFO] 'sg13g2_xor2_1'
[INFO] 'sg13g2_nor4_2'
[INFO] 'sg13g2_inv_2'
[INFO] 'sg13g2_nand2b_1'
[INFO] 'sg13g2_nand3_1'
[INFO] 'sg13g2_a22oi_1'
[INFO] 'sg13g2_a21oi_1'
[INFO] 'sg13g2_nor2_1'
[INFO] 'sg13g2_a21oi_2'
[INFO] 'sg13g2_xnor2_1'
[INFO] 'sg13g2_and2_1'
[INFO] 'sg13g2_nand4_1'
[INFO] 'sg13g2_nor3_1'
[INFO] 'sg13g2_inv_1'
[INFO] 'sg13g2_nand2_1'
[INFO] 'sg13g2_o21ai_1'
[INFO] 'sg13g2_or4_1'
[INFO] 'sg13g2_a221oi_1'
[INFO] 'sg13g2_nor4_1'
[INFO] 'sg13g2_a21o_1'
[INFO] 'sg13g2_and3_1'
[INFO] 'sg13g2_mux2_1'
[INFO] 'sg13g2_and4_1'
[INFO] 'sg13g2_nand2_2'
[INFO] 'sg13g2_and2_2'
[INFO] 'sg13g2_nand3b_1'
[INFO] 'sg13g2_or3_1'
[INFO] 'sg13g2_dfrbp_1'
[INFO] 'sg13g2_tiehi'
[INFO] 'sg13g2_buf_2'
[INFO] 'sg13g2_buf_4'
[INFO] 'sg13g2_buf_1'
[INFO] 'VIA_via5_6_5000_5000_5_5_840_840'
[INFO] 'VIA_via5_6_5000_1800_1_5_840_840'
[INFO] 'VIA_via5_6_2200_5000_5_2_840_840'
[INFO] 'VIA_via5_6_2200_1800_1_2_840_840'
[INFO] 'VIA_via4_5_5000_480_1_12_410_410'
[INFO] 'VIA_via3_4_5000_480_1_12_410_410'
[INFO] 'VIA_via2_3_5000_480_1_12_410_410'
[INFO] 'VIA_via1_2_5000_480_1_12_410_410'
[INFO] 'VIA_via4_5_2200_480_1_5_410_410'
[INFO] 'VIA_via3_4_2200_480_1_5_410_410'
[INFO] 'VIA_via2_3_2200_480_1_5_410_410'
[INFO] 'VIA_via1_2_2200_480_1_5_410_410'
[INFO] 'VIA_Via1_YX_so'
[INFO] 'VIA_Via2_YX_so'
[INFO] 'VIA_Via1_XXW_so'
[INFO] 'VIA_Via3_YX_so'
[INFO] 'VIA_Via2_YYS_so'
[INFO] Clearing cells...
[INFO] Merging GDS/OAS files...
	./platforms/ihp-sg13g2/gds/sg13g2_stdcell.gds
[INFO] Copying toplevel cell 'gcd'
INFO: Reading config file: ./platforms/ihp-sg13g2/fill.json
ERROR: Unrecognized fill:     - LAYER Metal2 RECT ( 74185 3780 ) ( 75185 4780 ) ;

  ./util/def2stream.py:79
  ./util/def2stream.py:143
Command exited with non-zero status 1
Elapsed time: 0:01.39[h:]min:sec. CPU time: user 1.22 sys 0.16 (99%). Peak memory: 402452KB.
```
The actual change is the following:
```
rect_pat = re.compile(r'''
  \s*\-\ LAYER\ (?P<layer>\S+)  # The layer name
  (?:                           # Non-capturing group
  \s+\+\ MASK\ (?P<mask>\d+)    # Mask, None if absent
  )?
  (?P<opc>                      # OPC, None if absent
  \s+\+\ OPC
  )?
  \s+RECT\s+
   \(\s+(?P<xlo>\d+)\s+(?P<ylo>\d+)\s+\)\s+   # rect lower-left pt
   \(\s+(?P<xhi>\d+)\s+(?P<yhi>\d+)\s+\)\s+ ; # rect upper-right pt
  ''',

```

**Comment by KrzysztofHerman** [comment]
Thank you for identifying issues related to the standard cells.  Actually we have some feedback from our users about these problems and we are working on that. I hope IHP platform available in ORFS will provide us more information.

File: flow/Makefile (hunk 2 above)
Comment (maliberty) [reply flow/Makefile]: You are still changing the default from nangate45/gcd to ihp-sg13g2/gcd

File: flow/designs/ihp-sg13g2/spi/config.mk (hunk 6 above)
Comment (maliberty) [reply flow/designs/ihp-sg13g2/spi/config.mk]: unresolved - it sounds like both lines can be removed.

**Comment by maliberty** [comment]
Some feedback has been addressed but some hasn't.

File: flow/Makefile [hunk 18]
```diff
@@ -97,12 +97,12 @@
...
 #DESIGN_CONFIG=./designs/ihp-sg13g2/aes/config.mk
 #DESIGN_CONFIG=./designs/ihp-sg13g2/ibex/config.mk
-DESIGN_CONFIG=./designs/ihp-sg13g2/gcd/config.mk
+#DESIGN_CONFIG=./designs/ihp-sg13g2/gcd/config.mk
 #DESIGN_CONFIG=./designs/ihp-sg13g2/spi/config.mk
 #DESIGN_CONFIG=./designs/ihp-sg13g2/riscv32i/config.mk
 
 # Default design
-#DESIGN_CONFIG ?= ./designs/ihp-sg13g2/gcd/config.mk
+DESIGN_CONFIG ?= ./designs/ihp-sg13g2/gcd/config.mk
```
Comment (rovinski) [inline flow/Makefile]: The default should be restored back to the original which is nangate45.

File: flow/designs/ihp-sg13g2/spi/config.mk (hunk 6 above)
Comment (KrzysztofHerman) [reply flow/designs/ihp-sg13g2/spi/config.mk]: done

File: flow/designs/ihp-sg13g2/spi/spi.sdc (hunk 7 above)
Comment (KrzysztofHerman) [reply flow/designs/ihp-sg13g2/spi/spi.sdc]: changed to constraint.sdc, spi.sdc removed

File: flow/designs/ihp-sg13g2/spi/src/spi.v
Comment (KrzysztofHerman) [reply flow/designs/ihp-sg13g2/spi/src/spi.v:1]: This is a basic SPI slave. We use it for our commercial flow. I added it here to compare the QoR between flows.

File: flow/platforms/ihp-sg13g2/config.mk (hunk 13 above)
Comment (KrzysztofHerman) [reply flow/platforms/ihp-sg13g2/config.mk:80]: No it does not.

File: flow/platforms/ihp-sg13g2/config.mk (hunk 12 above)
Comment (KrzysztofHerman) [reply flow/platforms/ihp-sg13g2/config.mk]: removed

File: flow/platforms/ihp-sg13g2/klayoutrc
Comment (KrzysztofHerman) [reply flow/platforms/ihp-sg13g2/klayoutrc:1]: removed

File: flow/platforms/ihp-sg13g2/cells_adders.v (hunk 8 above)
Comment (KrzysztofHerman) [reply flow/platforms/ihp-sg13g2/cells_adders.v]: Since we do not provide adders I removed the file cells_adders.v
It also was not referenced in the config.mk

File: flow/platforms/ihp-sg13g2/config.mk (hunk 11 above)
Comment (KrzysztofHerman) [reply flow/platforms/ihp-sg13g2/config.mk]: removed from DONT_USE_CELLS list

File: flow/Makefile (hunk 2 above)
Comment (KrzysztofHerman) [reply flow/Makefile]: Done

File: flow/platforms/ihp-sg13g2/cells_latch.v (hunk 9 above)
Comment (KrzysztofHerman) [reply flow/platforms/ihp-sg13g2/cells_latch.v]: I added mapping to IHP latch

File: flow/Makefile [hunk 19]
```diff
@@ -94,9 +94,15 @@
...
+
+#DESIGN_CONFIG=./designs/ihp-sg13g2/aes/config.mk
+#DESIGN_CONFIG=./designs/ihp-sg13g2/ibex/config.mk
+#DESIGN_CONFIG=./designs/ihp-sg13g2/gcd/config.mk
+#DESIGN_CONFIG=./designs/ihp-sg13g2/spi/config.mk
+#DESIGN_CONFIG=./designs/ihp-sg13g2/riscv32i/config.mk
+
 # Default design
-DESIGN_CONFIG ?= ./designs/nangate45/gcd/config.mk
+DESIGN_CONFIG=./designs/nangate45/gcd/config.mk
```
Comment (maliberty) [inline flow/Makefile]: It should be ?= not = so that it can be overridden.  The original line was
```
DESIGN_CONFIG ?= ./designs/nangate45/gcd/config.mk
```

**Review by maliberty — CHANGES_REQUESTED** [review]
Just one new item

**Comment by maliberty** [comment]
It would be good to add these designs into jenkins/* so they are run as part of the CI.

**Review by maliberty — APPROVED** [review]

**Comment by KrzysztofHerman** [comment]
Thank you for all comments. How can I integrate the designs with jenkins ?

**Comment by maliberty** [comment]
> Thank you for all comments. How can I integrate the designs with jenkins ?

At the top level there is a jenkins/ dir with three files.  In each we list the design run, eg https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/8a0f9b75f7a3aae76aba6071d37a7509ac18c6e7/jenkins/public_tests_all.Jenkinsfile#L31

You would add your designs in a similar style to the list that is already there for the other pdks.  I would add them to all three as your designs are all on the small side.

**Comment by KrzysztofHerman** [comment]
> > Thank you for all comments. How can I integrate the designs with jenkins ?
> 
> At the top level there is a jenkins/ dir with three files. In each we list the design run, eg
> 
> https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/8a0f9b75f7a3aae76aba6071d37a7509ac18c6e7/jenkins/public_tests_all.Jenkinsfile#L31
> 
> You would add your designs in a similar style to the list that is already there for the other pdks. I would add them to all three as your designs are all on the small side.

Ok. so let me do the changes.

File: jenkins/public_nightly.Jenkinsfile [hunk 20]
```diff
@@ -95,6 +95,11 @@ pipeline {
                    "jpeg gf180",
                    "riscv32i gf180",
                    "uart-blocks gf180";
+                   "aes ihp-sg13g2";
+                   "ibex ihp-sg13g2";
+                   "gcd ihp-sg13g2";
+                   "spi ihp-sg13g2";
+                   "riscv32i ihp-sg13g2";
```
Comment (maliberty) [inline jenkins/public_nightly.Jenkinsfile]: The list terminates with a ; the interior items need a ,

**Comment by maliberty** [comment]
I see in the CI that the new designs are failing the metrics checks in continuous-integration/jenkins/pr-head .  For example gcd
```
[ERROR] synth__design__instance__area__stdcell fail test: 5087.313 <= 2725.0
[ERROR] placeopt__design__instance__area fail test: 5419.61 <= 3022.0
[ERROR] detailedroute__route__wirelength fail test: 12007.0 <= 6849.0
[ERROR] finish__design__instance__area fail test: 24623.2 <= 3065.0

Failed metadata checks: 4 out of 22
```
I think you'll need to run these design, do 'make update_ok', and commit the revised metrics.

**Comment by maliberty** [comment]
Don't worry about the DependencyInstaller failures as that's unrelated.

**Comment by KrzysztofHerman** [comment]
> I see in the CI that the new designs are failing the metrics checks in continuous-integration/jenkins/pr-head . For example gcd
> 
> ```
> [ERROR] synth__design__instance__area__stdcell fail test: 5087.313 <= 2725.0
> [ERROR] placeopt__design__instance__area fail test: 5419.61 <= 3022.0
> [ERROR] detailedroute__route__wirelength fail test: 12007.0 <= 6849.0
> [ERROR] finish__design__instance__area fail test: 24623.2 <= 3065.0
> 
> Failed metadata checks: 4 out of 22
> ```
> 
> I think you'll need to run these design, do 'make update_ok', and commit the revised metrics.

I have run all designs and got clean runs. Unfortunately make update_ok returns an error:
```
make update_ok
[INFO][FLOW] Using platform directory /home/herman/asic/pdks/Github/kherman/OpenROAD-flow-scripts/flow/platforms/ihp-sg13g2
cp -f ./reports/ihp-sg13g2/gcd/base/metadata-base.json \
      ./designs/ihp-sg13g2/gcd//metadata-base-ok.json
cp: cannot stat './reports/ihp-sg13g2/gcd/base/metadata-base.json': No such file or directory
make: *** [/home/herman/asic/pdks/Github/kherman/OpenROAD-flow-scripts/flow/util/utils.mk:26: update_metadata] Error 1

``` 
and indeed the file `./reports/ihp-sg13g2/gcd/base/metadata-base.json` is not generated during the flow.

**Comment by maliberty** [comment]
Sorry, you'll need to run 'make metadata' first.  When I do that I see errors on gcd as well.

File: flow/Makefile [hunk 21]
```diff
@@ -94,9 +94,15 @@
...
+
+#DESIGN_CONFIG=./designs/ihp-sg13g2/aes/config.mk
+#DESIGN_CONFIG=./designs/ihp-sg13g2/ibex/config.mk
+#DESIGN_CONFIG=./designs/ihp-sg13g2/gcd/config.mk
+#DESIGN_CONFIG=./designs/ihp-sg13g2/spi/config.mk
+DESIGN_CONFIG=./designs/ihp-sg13g2/riscv32i/config.mk
+
 # Default design
-DESIGN_CONFIG ?= ./designs/nangate45/gcd/config.mk
+#DESIGN_CONFIG ?= ./designs/nangate45/gcd/config.mk
```
Comment (maliberty) [inline flow/Makefile]: You have again changed the default design.

Comment (maliberty) [reply flow/Makefile]: FYI, you can do ```DESIGN_CONFIG=/designs/ihp-sg13g2/riscv32i/config.mk make``` and avoid editing this file when switching designs.
